### PR TITLE
(WIP) Have min time interval for force compaction.

### PR DESCRIPTION
**Pull request by reddycharan**
Descriptions of the changes in this PR:

- minimum interval needed to trigger force compaction by LedgerDirsListener,
since the last main compaction. Otherwise if isForceGCAllowWhenNoSpace is enabled
and when disk reaches storage threshold, it will do busy force gc for every
'diskCheckInterval' (10 secs) untill the disk/bookie becomes writable again.
gc operation is an expensive operation and it puts load on metadataserver for
iterating over all ledgers metadata. So we should call forcegc judiciously when
disk/bookie reaches storage threshold.

- here main compaction will be Major compaction, if majorcompaction is
disabled/suspended then it will be minorcompaction. If both major and minor are
disabled/suspended then it will be just gc run.

- in LedgerDirsMonitor and in GarbageCollectorThread we should schedule tasks
with fixed delay instead of at fixed rate. This would ensure that there would be
uniform delay between executions.

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java [hunk 1]
```diff
@@ -297,7 +316,7 @@ public void start() {
         if (scheduledFuture != null) {
             scheduledFuture.cancel(false);
         }
-        scheduledFuture = gcExecutor.scheduleAtFixedRate(this, gcWaitTime, gcWaitTime, TimeUnit.MILLISECONDS);
+        scheduledFuture = gcExecutor.scheduleWithFixedDelay(this, gcWaitTime, gcWaitTime, TimeUnit.MILLISECONDS);
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: any reason why did you change this to `FixedDelay`?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java [hunk 2]
```diff
@@ -245,6 +250,20 @@ public void disableForceGC() {
         }
     }
 
+    public void enableForceGCIfMinIntervalElapsed() {
+        long curTime = System.currentTimeMillis();
+        if ((lastMainCompactionTime.longValue() - curTime) > forceCompactionInterval) {
```
Comment (jiazhai) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: Will `curTime` always bigger than `lastMainCompactionTime`?

Comment (jvrao) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: yeah curTime will be >= lastMain isn't it @reddycharan ?

Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: bummer..yes, it should be other way round. In next iteration I'm planning to add testcases.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java [hunk 3]
```diff
@@ -74,6 +74,8 @@
     final double majorCompactionThreshold;
     final long majorCompactionInterval;
     long lastMajorCompactionTime;
+    final long forceCompactionInterval;
+    final AtomicLong lastMainCompactionTime;
```
Comment (jvrao) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: why are we calling it 'main' compaction? Don't we need to delay as long as we run *any* compaction?

Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: @jvrao - here main compaction will be Major compaction, if majorcompaction is
  disabled/suspended then it will be minorcompaction. If both major and minor are
  disabled/suspended then it will be just gc run.

I dont want to delay if major compaction is enabled and just minor compaction has run recently. Because if forcegc is triggered immediately without delay, then there is good chance of reclaiming enough space to become writable again.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java [hunk 4]
```diff
@@ -359,8 +379,21 @@ public void runWithFlags(boolean force, boolean suspendMajor, boolean suspendMin
             minorCompacting.set(true);
             doCompactEntryLogs(minorCompactionThreshold);
             lastMinorCompactionTime = System.currentTimeMillis();
+            if (!enableMajorCompaction || suspendMajor) {
```
Comment (jvrao) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: Given the if condition above, we enter this block only if MajorCompaction is enabled right?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java [hunk 5]
```diff
@@ -359,8 +379,21 @@ public void runWithFlags(boolean force, boolean suspendMajor, boolean suspendMin
...
+                lastMainCompactionTime.set(lastMinorCompactionTime);
+            }
             gcStats.getMinorCompactionCounter().inc();
             minorCompacting.set(false);
+        } else if (((!enableMajorCompaction) || suspendMajor) && ((!enableMinorCompaction) || suspendMinor)) {
+            /*
+             * if both compactions (major and minor) are disabled / suspended,
+             * then consider regular gc run as main compaction
+             */
+            lastMainCompactionTime.set(curTime);
```
Comment (jvrao) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: What is the downside of any major/minor/gc as last compaction time? This  to make a decision if we need to run force GC or not right?

Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: I dont want to delay forcegc, if major compaction is enabled and just minor compaction has run recently. Because, if forcegc is triggered immediately without delay, then there is good chance of reclaiming enough space to become writable again.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java (hunk 1 above)
Comment (karanmehta93) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: @sijie `FixedDelay` would ensure that fixed amount of time passes between consecutive runs. This will prevent from GC process overburdening the bookie. With `FixedRate`, the duration of GC run is counted as part of the delay time.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java (hunk 3 above)
Comment (karanmehta93) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: yes @jvrao That's right.
I will rename this as `lastCompactionTime` and add a comment there.

**Comment by karanmehta93** [comment]
I will be taking over this PR. Major task is to add tests and fix the bug. 
@sijie @jvrao Any other concerns here before I start?

**Comment by karanmehta93** [comment]
run bookkeeper-server bookie tests

**Comment by eolivelli** [comment]
Are you using this patch in production at SF ?
cc @jvrao @karanmehta93

**Comment by karanmehta93** [comment]
@eolivelli We aren't using this patch in prod. While its a actual issue, the frequency of it in production is rare for us. Hence we didn't prioritize it.

**Comment by eolivelli** [comment]
I am closing this PR.

Feel free to open it again Charan if you want to complete the work.

**Comment by eolivelli** [comment]
closing in favour of #2094
